### PR TITLE
RM-7843 Replace Rhino.Mocks with Moq in Mixins

### PR DESCRIPTION
**Pull request by derlistigelurch**

### CODE REVIEW

File: Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs [hunk 1]
```diff
@@ -295,44 +294,43 @@ public void ClassContextInheritance_WithOverrides_FromSameConfiguration ()
...
       var suppressedMixins = new[] { typeof (double) };
-
-      var classBuilderMock = mockRepository.StrictMock<ClassContextBuilder> (builder, targetType);
-      var mixinBuilderMock = mockRepository.StrictMock<MixinContextBuilder> (classBuilderMock, mixinType, _mixinContextOrigin);
-
-      using (mockRepository.Ordered ())
-      {
-        builder
-            .Expect (
+
```
Comment (MichaelKetting) [inline Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs:303]:
```suggestion

```

File: Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs [hunk 2]
```diff
@@ -295,44 +294,43 @@ public void ClassContextInheritance_WithOverrides_FromSameConfiguration ()
...
-      var mixinBuilderMock = mockRepository.StrictMock<MixinContextBuilder> (classBuilderMock, mixinType, _mixinContextOrigin);
-
-      using (mockRepository.Ordered ())
-      {
-        builder
-            .Expect (
+    
+      var classBuilderMock = new Mock<ClassContextBuilder> (MockBehavior.Strict, builder.Object, targetType);
+      var mixinBuilderMock = new Mock<MixinContextBuilder> (MockBehavior.Strict, classBuilderMock.Object, mixinType, _mixinContextOrigin);
+
```
Comment (MichaelKetting) [inline Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs:306]:
```suggestion

```

File: Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs [hunk 3]
```diff
@@ -295,44 +294,43 @@ public void ClassContextInheritance_WithOverrides_FromSameConfiguration ()
...
-        builder
-            .Expect (
+    
+      var classBuilderMock = new Mock<ClassContextBuilder> (MockBehavior.Strict, builder.Object, targetType);
+      var mixinBuilderMock = new Mock<MixinContextBuilder> (MockBehavior.Strict, classBuilderMock.Object, mixinType, _mixinContextOrigin);
+    
+      var sequence = new MockSequence();
+    
+      builder
+            .Setup (
```
Comment (MichaelKetting) [inline Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs:310]:
```suggestion
      builder
            .InSequence (sequence)
            .Setup (
```

File: Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs [hunk 4]
```diff
@@ -342,90 +340,90 @@ public void AddMixinToClass_WithoutOrigin ()
...
```
Comment (MichaelKetting) [inline Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs:372]:
```suggestion
      builder
            .InSequence (sequence)
            .Setup (
```

File: Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs [hunk 5]
```diff
@@ -342,90 +340,90 @@ public void AddMixinToClass_WithoutOrigin ()
...
-      var mixinBuilderMock = mockRepository.StrictMock<MixinContextBuilder> (classBuilderMock, mixinType, _mixinContextOrigin);
-
-      using (mockRepository.Ordered ())
-      {
-        builder
-            .Expect (
+    
+      var classBuilderMock = new Mock<ClassContextBuilder> (MockBehavior.Strict, builder.Object, targetType);
+      var mixinBuilderMock = new Mock<MixinContextBuilder> (MockBehavior.Strict, classBuilderMock.Object, mixinType, _mixinContextOrigin);
+
```
Comment (MichaelKetting) [inline Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs:368]:
```suggestion

```

File: Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs [hunk 6]
```diff
@@ -342,90 +340,90 @@ public void AddMixinToClass_WithoutOrigin ()
...
       var suppressedMixins = new Type[0];
-
-      var classBuilderMock = mockRepository.StrictMock<ClassContextBuilder> (builder, targetType);
-      var mixinBuilderMock = mockRepository.StrictMock<MixinContextBuilder> (classBuilderMock, mixinType, _mixinContextOrigin);
-
-      using (mockRepository.Ordered ())
-      {
-        builder
-            .Expect (
+
```
Comment (MichaelKetting) [inline Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinConfigurationBuilderTest.cs:365]:
```suggestion

```

File: Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinContextBuilderTest.cs [hunk 7]
```diff
@@ -238,16 +231,15 @@ public void ReplaceMixin_SelfSuppressor_GenericDefinition ()
     public void ReplaceMixin_BaseSuppressor ()
     {
       _parentBuilderMock
-          .Expect (mock => mock.SuppressMixin (Arg<IMixinSuppressionRule>.Matches (
-              rule => ((MixinTreeReplacementSuppressionRule) rule).ReplacingMixinType == _mixinBuilder.MixinType
+          .Setup (mock => mock.SuppressMixin (It.Is<IMixinSuppressionRule> (              rule => ((MixinTreeReplacementSuppressionRule) rule).ReplacingMixinType == _mixinBuilder.MixinType
```
Comment (MichaelKetting) [inline Remotion/Mixins/UnitTests/Core/Context/FluentBuilders/MixinContextBuilderTest.cs:234]:
```suggestion
          .Setup (mock => mock.SuppressMixin (It.Is<IMixinSuppressionRule> (
              rule => ((MixinTreeReplacementSuppressionRule) rule).ReplacingMixinType == _mixinBuilder.MixinType
```

File: Remotion/Mixins/UnitTests/Core/ExtendsAttributeTest.cs [hunk 8]
```diff
@@ -274,13 +271,11 @@ public void Apply_InvalidOperation ()
       var attribute = new ExtendsAttribute (typeof (object));
 
       _configurationBuilderMock
-          .Expect (mock => mock.AddMixinToClass (MixinKind.Extending, null, null, MemberVisibility.Private, null, null, null))
-          .IgnoreArguments()
-          .Throw (new InvalidOperationException ("Foofa."));
+.Setup  (mock => mock.AddMixinToClass (It.IsAny<MixinKind>(),It.IsAny<Type>(),It.IsAny<Type>(),It.IsAny<MemberVisibility>(),It.IsAny<IEnumerable<Type>>(),It.IsAny<IEnumerable<Type>>(),It.IsAny<MixinContextOrigin>())).Throws  (new InvalidOperationException ("Foofa."))
+.Verifiable();
```
Comment (MichaelKetting) [inline Remotion/Mixins/UnitTests/Core/ExtendsAttributeTest.cs:275]: Please fix alignment

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
The RhinoMocks dependency removal in csproj is still open, too.

File: Remotion/Mixins/UnitTests/Core/CodeGeneration/ObjectFactoryImplementationTest.cs [hunk 9]
```diff
@@ -64,16 +65,17 @@ public void CreateInstance_WithNoPreparedMixinInstances_SetsUpEmptyPreparedMixin
     {
       object[] actualSuppliedInstances = null;
       _defaultPipelineMock
-          .Expect (mock => mock.Create (typeof (BaseType1), ParamList.Empty, allowNonPublicConstructor: false))
-          .Return (new object ())
-          .WhenCalled (mi => actualSuppliedInstances = MixedObjectInstantiationScope.Current.SuppliedMixinInstances);
+          .Setup (mock => mock.Create (typeof (BaseType1), ParamList.Empty, false))
```
[the remark this answers is not in the document]
Comment (derlistigelurch) [reply Remotion/Mixins/UnitTests/Core/CodeGeneration/ObjectFactoryImplementationTest.cs:68]: Sadly it's not possible to do that. Inserting the named argument results in the following error message
`An expression tree may not contain a named argument specification`.

File: Remotion/Development/RhinoMocks/UnitTesting/ClassWithSerializationCallbacksBase.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Runtime.Serialization;
+
+namespace Remotion.Development.UnitTesting
```
Comment (MichaelKetting) [inline Remotion/Development/RhinoMocks/UnitTesting/ClassWithSerializationCallbacksBase.cs:20]: Can remain in Development.Core. Development.RhinoMocks and Devleopment.Moq depend on Development.Core

File: Remotion/Development/RhinoMocks/Development.RhinoMocks.csproj [hunk 11]
```diff
@@ -14,13 +14,15 @@
     <Compile Include="..\..\..\SharedSource\Core\FunctionalProgramming\*.cs" Link="SharedSource\FunctionalProgramming\%(Filename)%(Extension)" />
     <Compile Include="..\..\..\SharedSource\Core\Utilities\*.cs" Link="SharedSource\Utilities\%(Filename)%(Extension).cs" />
     <Compile Include="..\..\..\SharedSource\Development\RhinoMocks\UnitTesting\**\*.cs" Link="SharedSource\RhinoMocks\UnitTesting\%(RecursiveDir)%(Filename)%(Extension).cs" />
-    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\PrivateInvoke.cs" Link="SharedSource\UnitTesting\PrivateInvoke.cs" />
-    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\ThreadRunner.cs" Link="SharedSource\UnitTesting\ThreadRunner.cs" />
     <Compile Include="..\..\..\SharedSource\Core\RoslynAnnotations\NullableAnnotations.cs" Link="SharedSource\RoslynAnnotations\NullableAnnotations.cs" />
+    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\Serializer.cs" Link="SharedSource\Development\UnitTesting\Serializer.cs" />
```
Comment (MichaelKetting) [inline Remotion/Development/RhinoMocks/Development.RhinoMocks.csproj:18]: Drop. Should be provided via Development.Core dependency.
```suggestion
```

File: Remotion/Development/Moq/UnitTesting/Serializer.Partial.cs [hunk 12]
```diff
@@ -18,7 +18,7 @@
 
 namespace Remotion.Development.UnitTesting
 {
-  public partial class Serializer
+  public static partial class Serializer
```
Comment (MichaelKetting) [inline Remotion/Development/Moq/UnitTesting/Serializer.Partial.cs:21]: Drop. Development.Moq already references Development.Core, that way, it gets access to the Serializer type needed by SerializationCallbackTester.

File: Remotion/Development/Moq/Development.Moq.csproj [hunk 13]
```diff
@@ -14,13 +14,15 @@
     <Compile Include="..\..\..\SharedSource\Core\FunctionalProgramming\*.cs" Link="SharedSource\FunctionalProgramming\%(Filename)%(Extension)" />
     <Compile Include="..\..\..\SharedSource\Core\Utilities\*.cs" Link="SharedSource\Utilities\%(Filename)%(Extension).cs" />
     <Compile Include="..\..\..\SharedSource\Development\Moq\UnitTesting\**\*.cs" Link="SharedSource\Moq\UnitTesting\%(RecursiveDir)%(Filename)%(Extension).cs" />
-    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\PrivateInvoke.cs" Link="SharedSource\UnitTesting\PrivateInvoke.cs" />
-    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\ThreadRunner.cs" Link="SharedSource\UnitTesting\ThreadRunner.cs" />
     <Compile Include="..\..\..\SharedSource\Core\RoslynAnnotations\NullableAnnotations.cs" Link="SharedSource\RoslynAnnotations\NullableAnnotations.cs" />
+    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\Serializer.cs" Link="SharedSource\Development\UnitTesting\Serializer.cs" />
```
Comment (MichaelKetting) [inline Remotion/Development/Moq/Development.Moq.csproj:18]:
```suggestion
```

File: Remotion/Development/Mixins/Development.Mixins.csproj [hunk 14]
```diff
@@ -15,6 +15,7 @@
     <ProjectReference Include="..\..\Core\Tools\Core.Tools.csproj" />
     <ProjectReference Include="..\..\Mixins\Core\Mixins.Core.csproj" />
     <ProjectReference Include="..\Core\Development.Core.csproj" />
+    <ProjectReference Include="..\Moq\Development.Moq.csproj" PrivateAssets="All"/>
```
Comment (MichaelKetting) [inline Remotion/Development/Mixins/Development.Mixins.csproj:18]: Drop.
```suggestion
```

This dependency should not be needed. Otherwise, we introduce a dependency on a particular mocking framework into a general purpose assembly.

File: Remotion/Development/Core/Development.Core.csproj [hunk 15]
```diff
@@ -12,9 +12,7 @@
     <RootNamespace>$(AssemblyName)</RootNamespace>
   </PropertyGroup>
   <ItemGroup>
-    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\**\*"
-             Link="SharedSource\UnitTesting\%(RecursiveDir)%(Filename)%(Extension)"
-             Exclude="**\PEVerifyVersion.cs;**\NUnit\**" />
+    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\**\*" Link="SharedSource\UnitTesting\%(RecursiveDir)%(Filename)%(Extension)" Exclude="**\PEVerifyVersion.cs;**\NUnit\**;**/Serializer.cs" />
```
Comment (MichaelKetting) [inline Remotion/Development/Core/Development.Core.csproj:15]: revert

File: SharedSource/Development/UnitTesting/Serializer.cs [hunk 16]
```diff
@@ -27,7 +27,7 @@ namespace Remotion.Development.UnitTesting
   /// Provides quick serialization and deserialization functionality for unit tests.
   /// </summary>
   /// <remarks>The methods of this class use a <see cref="BinaryFormatter"/> for serialization.</remarks>
-  public static partial class Serializer
+  static partial class Serializer
```
Comment (MichaelKetting) [inline SharedSource/Development/UnitTesting/Serializer.cs:30]: Good fix. Should be a dedicated commit since it's unrelated to the current task.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
RhinoMocks: SerializationCallbackTester and IMockRepository should get a namespace update.
RhinoMocks and Moq: ClassWithSerializationCallbacksBase should remain in Development.Core

File: Remotion/Development/Core/Development.Core.csproj [hunk 17]
```diff
@@ -12,9 +12,7 @@
     <RootNamespace>$(AssemblyName)</RootNamespace>
   </PropertyGroup>
   <ItemGroup>
-    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\**\*"
-             Link="SharedSource\UnitTesting\%(RecursiveDir)%(Filename)%(Extension)"
-             Exclude="**\PEVerifyVersion.cs;**\NUnit\**" />
+    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\**\*" Link="SharedSource\UnitTesting\%(RecursiveDir)%(Filename)%(Extension)" Exclude="**\PEVerifyVersion.cs;**\NUnit\**" />
```
Comment (MichaelKetting) [inline Remotion/Development/Core/Development.Core.csproj:15]: Please revert format change. Seems an auto-format kicked in?

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
a format revert is missing, otherwise looking good.

**Review by MichaelKetting — APPROVED** [review]
changeset looks good. ready for rebase/flatten
